add better route parsing

trvswgnr opened the pull request:
`¯\_(ツ)_/¯`

cod1r commented:
This PR is too good.

trvswgnr commented:
@twny fixed the conflicts! i also moved the hashmap into the main function so it doesn't get redefined and did a little housekeeping

twny commented on src/main.rs:
```diff
@@ -1,80 +1,217 @@
 
-const body_404: &str = include_str!("404.html");
-const body_index: &str = include_str!("index.html");
+const BODY_404: &str = include_str!("404.html");
+const BODY_INDEX: &str = include_str!("index.html");
```
good catch constants should be all caps

twny commented on src/main.rs:
```diff
@@ -1,80 +1,217 @@
 
-fn get_foo(_: &str) -> String {
+fn handler_foo(_: &str) -> String {
     let status = "HTTP/1.1 404 Not Found \r\n";
     let body = "<html><h1>Foo</h1></html>";
     let size = format!("Content-Length: {}\r\n", body.len());
     let response = format!("{status}{size}\r\n{body}");
-    return response;
+
+    response
 }
```
i don't think we need this code anymore

trvswgnr commented:
- resolved conflicts with most recent changes
- updated to use `Arc` for sharing `HashMap` between threads without creating a new object every time
- changed hashmap to use `&str` instead of String

twny commented:
this is so legit! i can't wait to review this on stream tomorrow and merge it. then we can add thread pools !

twny commented on src/main.rs:
```diff
@@ -40,44 +42,177 @@ fn main() {
         // TODO make this a thead pool
-        thread::spawn(|| {
-            handle_connection(stream);
+        thread::spawn(move || {
```
```suggestion
        thread::spawn(|| {
```

twny commented on src/main.rs:
```diff
@@ -40,44 +42,177 @@ fn main() {
+    let query = route_parts.next().unwrap_or("");
+    let mut headers = HashMap::new();
+
+    for (index, header) in request.iter().enumerate() {
```
```suggestion
    for (index, header) in request.iter().enumerate().skip(1) {
```

twny commented on src/main.rs:
```diff
@@ -40,44 +42,177 @@ fn main() {
+    let mut headers = HashMap::new();
+
+    for (index, header) in request.iter().enumerate() {
+        if index > 0 {
```
```suggestion
```

twny commented on src/main.rs:
```diff
@@ -40,44 +42,177 @@ fn main() {
+
+    for (index, header) in request.iter().enumerate() {
+        if index > 0 {
+            if !header.contains(':') || header.starts_with('{') {
+                continue;
+            }
+            let mut split = header.split(':');
+            headers.insert(
+                // key
+                split.next().unwrap().trim().to_string(),
+                // value
+                split.collect::<Vec<&str>>().join(":").trim().to_string(),
+            );
+        }
```
```suggestion
```

twny replied:
```suggestion
        if !header.contains(':') || header.starts_with('{') {
            continue;
        }
        let mut split = header.split(':');
        headers.insert(
            // key
            split.next().unwrap().trim().to_string(),
            // value
            split.collect::<Vec<&str>>().join(":").trim().to_string(),
        );
    }
```
